### ISSUE TITLE
Make token URL clearer when no-browser is given

### ISSUE THREAD

**Issue by Carreau** [jupyter/notebook#1980]
Or be more explicit on the login webpage that a token need to be passed, 
and can be found in the Cli that launched the notebook. 

Otherwise users can't log-in after upgrading to 4.3

(and print the token on the CLI in a more visible manner.)

**Comment by takluyver** [comment]
I think the page should be clearer about why it's prompting for a password (this was already suggested on one of the issues), but I don't think we should remove the `--no-browser` option. It still does something useful, and there are ways to get a token other than letting it start a browser.

**Comment by Carreau** [comment]
Well, maybe not removing but be clearer. If you use --no-browser, then you have next to 0 chances to have access to the token.

**Comment by takluyver** [comment]
It should still print a URL containing the token in the terminal, and you can use `jupyter notebook list` to get it. If you're using a script to access it, it can get it from the JSON file with the notebook server info, I believe.

**Comment by Carreau** [comment]
@fperez had more issue with this. In particular the cookie is not set on both localhost and 127.0.0.1 which is confising. 

Users using anaconda may not even see the token in the terminal. 

The token should be logged in the terminal when the login page is hit (for user to not have to scroll).

**Comment by fperez** [comment]
I have to say, the current usability of this for new comers is basically -infinity.  It nearly stopped me dead on my tracks, and I know a bit more than many of our new users.  For someone who doesn't understand all the moving parts, this can be a complete show-stopper right now.

I also understand the important security reasons why this is present, not trying to discount that.  But we need to figure out an improvement for the UX.

BTW, in my case part of the issue is that I seem to have `no-browser` set somewhere, but even with @Carreau's help we couldn't figure out where.  `jupyter notebook` in my case doesn't trigger a browser, but I can't find any customization to that effect anywhere.

And having `jupyter notebook list` be the solution isn't going to help new users: many of them won't even know this subcommand exists (I'd completely forgotten it).

**Comment by takluyver** [comment]
I don't think it's so bad for newcomers - the default experience is with it opening a browser directly on launch, which works (as far as we know). The problems are mainly for existing users who already have a different way of doing things which doesn't work with this. Like yourself and the hidden no-browser setting.

The login page (which you see if you miss the token) will point users to `jupyter notebook list` as of 4.3.1.

The only other thing I can think of to do is to special case connections which say they're coming from `localhost` or `127.0.0.1` (and the correct port). I.e. check the origin against what the server expects it to be, rather than comparing two values sent by the browser. But I'm not wild on this, because:

- Adding an exception to any security mechanism is a risk, because someone might find a way to exploit that exception. The flaw that prompted adding this mechanism highlighted just how complex browsers and web security are. More special cases makes it harder to be sure we're doing things right.
- Special casing specific origins means more surprising behaviour when you're legitimately accessing it at a different domain.

> the cookie is not set on both localhost and 127.0.0.1 which is confising.

I understand that it's confusing, but from the browser's point of view, those are different domains. We [can't set a cookie for another domain](http://stackoverflow.com/a/6761443/434217).

**Comment by takluyver** [comment]
(Looking at that last link more carefully, I see that it may actually be possible to set the cookie across both domains - I posted it only having read the first line ;-)

**Comment by mlgill** [comment]
Hi all, 

Was just surprised by this issue while building a Docker container that I was testing on an AWS instance (i.e. no desktop). 

Based on my reading of the [instructions](http://jupyter-notebook.readthedocs.io/en/latest/security.html), I tried both of the following launch commands with notebook v4.3:

```bash
jupyter-notebook --no-browser --port 8888 --ip=0.0.0.0 \
--config NotebookApp.password='' \
--config NotebookApp.token='' \
--config NotebookApp.password_required=False
```

and

```bash
jupyter-notebook --port 8888 --ip=0.0.0.0
```

However neither of these launch commands successfully removed the password requirement within the Docker container. 

I was only able to remove it by forcing the notebook version to be 4.2.3. This is fine in the short term, but not a long term solution, obviously.

If I have missed something or you have suggestions, let me know. Thanks.

**Comment by takluyver** [comment]
@mlgill : the way you're trying to pass config in does not work:

```
# Wrong:
--config NotebookApp.token=''

# Right:
--NotebookApp.token=''
```

The `--config` option is for specifying the path to a config file.

**Comment by takluyver** [comment]
I should also point out that disabling all security and then listening on all interfaces is not recommended (to put it mildly). Hopefully you already know that and are just testing, but I wanted to make sure it's clear for anyone else reading this thread.

The docs have more info on how to [secure a server that will run publicly](http://jupyter-notebook.readthedocs.io/en/latest/public_server.html#securing-a-notebook-server).

**Comment by minrk** [comment]
I've retitled the issue to point to making the info clearer, since we shouldn't disable token auth, or even recommend doing so.

@mlgill when you launch the notebook server, the token is logged, and you can use this to connect. Or, if the token being generated causes inconvenience, you can use a predetermined value, e.g.

    jupyter notebook --NotebookApp.token='super-secret'

#1972 should have already covered most of this. The login page looks like this now:

<img width="633" alt="screen shot 2016-12-13 at 13 19 49" src="https://cloud.githubusercontent.com/assets/151929/21315421/64bb29d4-c5fc-11e6-9393-956bea2d8f51.png">

which better explains why the login page is shown, and shows in detail the exact steps to get the information you need. Logging in with a token is needed once per browser+port+machine combo, so once a user has done this one time on their laptop, they should never need to do it again.

#1994 should fill in the one remaining gap I see of extra emphasis of the information already in the log output, which will now look like:

```
[C 15:17:09.076 NotebookApp]

    Copy/paste this URL into your browser when you connect for the first time,
    to login with a token:
        http://localhost:8888/?token=d0b6a3e08ee7d81ce602e8e5f5f3bd4d37f4380f36798df7
```

Any further feedback on what we should do would be greatly appreciated.

I'm sorry for the consternation this has caused. I'd like to get 4.3.1 out very soon with these improvements ASAP.

As a side issue, @fperez, I'm not sure how you would have `open_browser = False` if it's not in your `jupyter_notebook_config.py`, but it could happen without any configuration if the webbrowser module cannot find any browser (in which case a `No web browser found:` message is logged).

**Comment by roycoding** [comment]
I was affected here in part because I had cookies disabled. It might be worthwhile to explicitly mention on the login page that cookies must be enabled (specifically for 127.0.0.1 and localhost).

**Comment by minrk** [comment]
@roycoding good point!

**Comment by fperez** [comment]
@minrk, what are all the locations where `jupyter_notebook_config.py` might be read from? I can't seem to find that in `~/Library/Jupyter` nor in `~/.ipython`, nor does a Spotlight search find any occurrence under my home dir...

Happy to help track this down, if nothing else, so we understand what's going on.

**Comment by fperez** [comment]
@takluyver, may I suggest that when a user (me, in this case :) reports that they are having a real usability struggle with our system, replying "the problems you think you're having aren't as bad as you say they are" is not the ideal response?

There may be any number of reasons why users don't get the token info, they lose their browser, their notebook was launched externally, etc.  As it currently stands, there's a very good chance they are going to be completely stumped by a login page they'd never, ever seen before, and for which they have no idea what to do with.

Let's instead figure out a solution that improves the situation.  I know that the intersection of good security and good UX is often the empty set, but let's see what we can do here...

**Comment by michaelpacer** [comment]
@fperez I actually needed to use this regarding the locations where` jupyter_notebook_config.py` in tracking down some remnant issues with a bad install of `jupyter_contrib_nbextensions`.

The command you want is

    jupyter --paths

`jupyter --config` only returns the most prominent config option. 

I actually was just trying to think about how to pass the json version of the output `juptyer --paths --json` to be parsed, extract the list of paths and pass that to a grep (in my case for `jupyter_contrib_nbextensions`). 

But after hacking at it for a while, I'm pretty sure that's way too much to be happening in sequence of pipes so I should just write a python script for it. It'd probably be helpful for these kinds of cases where some setting exists somewhere on a user's machine and we don't know where the setting is coming from.

**Comment by jeremyherbert** [comment]
Has the `password_required` listed here: http://jupyter-notebook.readthedocs.io/en/latest/config.html

been removed? My latest pip installed jupyter says the following:

```
[W 06:14:52.969 NotebookApp] Config option `password_required` not recognized by `NotebookApp`.  Did you mean `password`?
```

**Comment by fperez** [comment]
Thanks @michaelpacer! That was useful, now I know for the future...

Unfortunately, I didn't find any browser-related config flags in any of the paths I got for config:

```
denali[~]> jupyter --paths
config:
    /Users/fperez/.jupyter
    /Users/fperez/usr/conda/etc/jupyter
    /usr/local/etc/jupyter
    /etc/jupyter
```

But this rubber-ducking helped me!! Given that there was nothing left, I scoured my environment, and found this:

```
denali[~]> env | grep -i brow
BROWSER=google-chrome&
```

This is a carryover of my `bashrc` from my Ubuntu days, but on OSX calling `google-chrome` doesn't do anything.  So what was happening in my case was that the browser command was failing, leaving me without a browser.

This shows another failure mode that can affect an unsuspecting user: if for whatever reason they don't get the browser to open, they're left to read startup logs (something a new user isn't necessarily always adept at).

I recently had the opportunity to help a user who was a complete newbie to Python/Jupyter, and it was really satisfactory to be able to get them to go from zero to working notebook with plots in the course of a short session (and it got them excited about the possibilities).  But this person had never interacted with a terminal at all nor had any real experience programming. I'm 100% certain that "read the log until you find the token URL, then use that to load the session in a browser" would be simply a bridge too far for them.

**Comment by takluyver** [comment]
@fperez sorry if my response was curt. This is a classic bug that's really annoying for the people it affects, but I still maintain that the vast majority of the new users you're concerned about will not be affected. Obviously we will take any bug you file fairly seriously however it's worded, but I find your bug reports sometimes have a little bit of the "this must be fixed now" flavour. And I find that all the more surprising because you have been the one reading and responding to plenty of bug reports in the past, so you know what makes a good one from the maintainers' point of view.

Back to the issue:

* To tackle the specific issue that caused your problem, maybe if $BROWSER is set and launching a browser fails, we should assume it's set badly, and try again as if it wasn't set. Scanning through the code of the `webbrowser` module, there doesn't seem to be a good way to do that, but if we think it will affect enough users, I'm sure we can come up with something.
* I totally agree that it's a poor user experience if the server doesn't launch the browser. But I don't see a great way to improve this in general. AIUI, we need to get a secret to the browser. If we can launch the browser, we do that via the command line. If not, we make it available in the terminal and on the filesystem for the user to pick up. We can't transfer it over HTTP (unless I'm missing a trick), because that's the channel we're trying to secure. We can definitely make it clearer what's going on, which we're doing in 4.3.1, but I don't see a way to avoid it.
* If we can't launch the browser and the user isn't looking at a terminal (e.g. because they've launched the notebook from a GUI), we don't have any obvious way to communicate with them even to point out what's going wrong. We could try to find an available GUI toolkit (tkinter is usually there) and open a window with information. This feels very much like a band-aid solution, though.
* We could try to push users into setting a password, so that they know what to enter on the login page. E.g. the Anaconda launcher could get the user to set a password when they first launch the notebook. But this isn't great UX either, and it's extra work for less security (human chosen passwords are easier to guess than a generated token).
* We could provide a GUI [notebook server manager](https://github.com/takluyver/nbmanager) as part of a default install, and add options to launch a browser and to copy the URL from that interface. That gets us back into the local GUI development we've avoided by making a web interface, but maybe that's OK for a tool with much more limited scope.

**Comment by minrk** [comment]
@fperez I'm fully on board, and trying to get out 4.3.1 with improvements ASAP, before this bites too many more people. I'd like to do it by the end of today, if possible, but tomorrow seems more likely. What do you see as missing in what we already have in master (plus #1994)? Right now, we have:

- complete information on the login page about why the page is shown and how to get the information needed (see above screenshot)
- more emphasized information in the log output, for users who don't launch a browser (#1994)
- detailed explanation of what's going on [in the docs](http://jupyter-notebook.readthedocs.io/en/latest/security.html)

> Has the password_required listed here: http://jupyter-notebook.readthedocs.io/en/latest/config.html been removed?

@jeremyherbert  I don't believe `password_required` has been backported to 4.x. It's still in master, awaiting 5.0 release.

**Comment by fperez** [comment]
@takluyver, I do apologize if I have made it sound like "must be done now", that's certainly not my intent.  B/c of being overloaded on many fronts, the reality is that lately I tend to file bugs only that i think hit us pretty hard, but perhaps my wording was sub-optimal.

I wasn't even worried about myself (I can work my way out of the situation), but more about the impact it can have on new users: while I think you're right in that it will be a minority, it can still be severe for those it hits.

And it's worth seeing how this can happen in unexpected ways:  I was just now working with a colleague who was stumped by this after a conda update.  It turns out that his default browser on OSX is Safari, so after starting `jupyter notebook`, Safari opens for him.  But for Jupyter he prefers Chrome, so he has the reflex to immediately kill/hide Safari and go over to `localhost:8888` on Chrome.

Well, guess what :)  Now, this puts him on a weird login/password page he'd never seen before, and was stumped.

Fortunately he'd just done the conda update with me in the room and I saw what happened, and was able to help him out.  But it shows that there can be multiple ways in which users can find themselves in this situation that we didn't necessarily foresee...

In any case, given that 4.3.1 is near-ready as per @minrk's description above, that sounds like a good amount of info/explanation for most cases.  We can keep collecting feedback on how this impacts the broader user base, and refine our approach.

But no, my bugs aren't "special", sorry again if it seemed that way.

**Comment by minrk** [comment]
A summary of the users who are likely to face the token auth, where the default open-with-token isn't quite magical enough:

- 'notebook' browser isn't default browser, or other browser hopping (e.g. switching around for testing)
- run with `--no-browser`, or browser launching doesn't work / isn't available (e.g. remote notebooks, docker, etc.)
- use private browsing, or other pattern where cookies are cleared often

In each of these cases, the token should only be needed for the first time a browser connects to a notebook server on a given ip/port, as a cookie is used thereafter, and cookies remain valid across notebook restarts. It should be considered an 'initial setup' step, rather than an ongoing interaction with the notebook server every time you launch. Resetting cookies will require logging in again. While *browsers* share cookies across ports, we specifically don't use the same cookie for each host:port combination, because different notebook configurations could cause conflicts on the cookie secret. We could consider reversing that, as I think the experience for *most* cases would be better (i.e. all localhost notebooks are actually local). It could cause surprising logout events for anyone who tunnels localhost to remote notebooks, though, or uses temporary config with a fresh cookie secret for testing.

**Comment by fperez** [comment]
Thanks @minrk and team for all the work and thought on this one, I know it sits at a very uncomfortable intersection of security and UX.

One last thought (possibly for 4.3.2, given the release timeline): should we offer a `jupyter password` command that does the whole pwd creation/hashing/writing dance in one step at the cmd line? 

In the past, needing a password was relatively rare, so having the process be [what explain here](http://jupyter-notebook.readthedocs.io/en/latest/public_server.html) seemed sufficient.  But if more and more users are likely to need this, we might want to streamline it as much as possible...

Thoughts?

**Comment by pymilo** [comment]
I was having the same issue after updating Anaconda to the version 1.6.0. I have configured my mac to run Google-Chrome as the default browser but any reason running > jupyter notebook was not working. I solved by uncommenting and editing this line 

c.NotebookApp.browser = u'open -a /Applications/Google\ Chrome.app %s'

in the ~/.jupyter/jupyter_notebook_config.py file. Hope thsi is useful for someone having the same issue.
